staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/controller.go

```diff
@@ -92,31 +93,55 @@ func (c *Controller) Run(staticSpec *spec.Swagger, openAPIService *handler.OpenA
 		return
 	}
 
-	// create initial spec to avoid merging once per CRD on startup
-	crds, err := c.crdLister.List(labels.Everything())
-	if err != nil {
-		utilruntime.HandleError(fmt.Errorf("failed to initially list all CRDs: %v", err))
-		return
+	var stopWorkerCh chan struct{}
+	var stoppedWorker chan struct{}
+	sleepControl := &helper.SleepControl{
+		WakeupFn: func() {
+			// create initial spec to avoid merging once per CRD on startup
+			crds, err := c.crdLister.List(labels.Everything())
+			if err != nil {
+				utilruntime.HandleError(fmt.Errorf("failed to initially list all CRDs: %v", err))
+				return
+			}
+			for _, crd := range crds {
+				if !apiextensionshelpers.IsCRDConditionTrue(crd, apiextensionsv1.Established) {
+					continue
+				}
+				newSpecs, changed, err := buildVersionSpecs(crd, nil)
+				if err != nil {
+					utilruntime.HandleError(fmt.Errorf("failed to build OpenAPI spec of CRD %s: %v", crd.Name, err))
+				} else if !changed {
+					continue
+				}
+				c.crdSpecs[crd.Name] = newSpecs
+			}
+			if err := c.updateSpecLocked(); err != nil {
+				utilruntime.HandleError(fmt.Errorf("failed to initially create OpenAPI spec for CRDs: %v", err))
+				return
+			}
+
+			// only start one worker thread since its a slow moving API
+			klog.V(4).Infof("Starting OpenAPI v2 controller for CRDs")
+			stopWorkerCh = make(chan struct{})
+			stoppedWorker = make(chan struct{})
+			go func(stopWorkerCh <-chan struct{}, stoppedWorker chan<- struct{}) {
+				defer close(stoppedWorker)
+				wait.Until(c.runWorker, time.Second, stopWorkerCh)
+			}(stopWorkerCh, stoppedWorker)
+		},
+		SleepFn: func() {
+			if stopWorkerCh != nil {
+				klog.V(4).Infof("Stopping OpenAPI v2 controller for CRDs")
+				close(stopWorkerCh)
+				<-stoppedWorker
+				stopWorkerCh = nil
+				stoppedWorker = nil
+			}
+		},
+		FallAsleepAfter: time.Minute,
 	}
-	for _, crd := range crds {
-		if !apiextensionshelpers.IsCRDConditionTrue(crd, apiextensionsv1.Established) {
-			continue
-		}
-		newSpecs, changed, err := buildVersionSpecs(crd, nil)
-		if err != nil {
-			utilruntime.HandleError(fmt.Errorf("failed to build OpenAPI spec of CRD %s: %v", crd.Name, err))
-		} else if !changed {
-			continue
-		}
-		c.crdSpecs[crd.Name] = newSpecs
-	}
-	if err := c.updateSpecLocked(); err != nil {
-		utilruntime.HandleError(fmt.Errorf("failed to initially create OpenAPI spec for CRDs: %v", err))
-		return
-	}
-
-	// only start one worker thread since its a slow moving API
-	go wait.Until(c.runWorker, time.Second, stopCh)
+	openAPIService.RegisterSynchronousHandlerCallback(sleepControl.Tick)
```

Comment:
Is this function defined in this PR?

Reply:
No, it's in kube-openapi:
```diff
diff --git a/pkg/handler/handler.go b/pkg/handler/handler.go
index 37cb96f..978c0b2 100644
--- a/pkg/handler/handler.go
+++ b/pkg/handler/handler.go
@@ -63,6 +63,9 @@ type OpenAPIService struct {
    specCache  cached.Replaceable[*spec.Swagger]
    jsonCache  cached.Data[timedSpec]
    protoCache cached.Data[timedSpec]
+
+   lock      sync.RWMutex
+   callbacks []func()
 }

 // NewOpenAPIService builds an OpenAPIService starting with the given spec.
@@ -108,6 +111,15 @@ func (o *OpenAPIService) UpdateSpecLazy(swagger cached.Data[*spec.Swagger]) {
    o.specCache.Replace(swagger)
 }

+// RegisterSynchronousHandlerCallback registers a callback to be called at the beginning of the handler.
+// It can be used to prime the spec.
+func (o *OpenAPIService) RegisterSynchronousHandlerCallback(callback func()) {
+   o.lock.Lock()
+   defer o.lock.Unlock()
+
+   o.callbacks = append(o.callbacks, callback)
+}
+
 func ToProtoBinary(json []byte) ([]byte, error) {
    document, err := openapi_v2.ParseDocument(json)
    if err != nil {
@@ -156,6 +168,14 @@ func (o *OpenAPIService) RegisterOpenAPIVersionedService(servePath string, handl
                    if clause.SubType != accepts.SubType && clause.SubType != "*" {
                        continue
                    }
+
+                   // call callback to potentially prime the spec.
+                   o.lock.RLock()
+                   defer o.lock.RUnlock()
+                   for _, callback := range o.callbacks {
+                       callback()
+                   }
+
                    // serve the first matching media type in the sorted clause list
                    mutex.Lock()
                    result := accepts.GetDataAndEtag.Get()
```